runtime/lua/vim/lsp/_completion.lua

```diff
@@ -273,4 +381,396 @@ function M.omnifunc(findstart, base)
+  cursor_row = cursor_row - 1
+  local bufnr = api.nvim_get_current_buf()
+  local expand_snippet = completion_item.insertTextFormat == protocol.InsertTextFormat.Snippet
+    and Context.expand_snippet
+    and (completion_item.textEdit ~= nil or completion_item.insertText ~= nil)
+  local resolve_provider = (client.server_capabilities.completionProvider or {}).resolveProvider
+
+  local function clear_word()
+    if not expand_snippet then
+      return nil
+    end
+
+    -- Remove the already inserted word.
+    local start_char = cursor_col - #completed_item.word
+    local line = api.nvim_buf_get_lines(bufnr, cursor_row, cursor_row + 1, true)[1]
+    api.nvim_buf_set_text(bufnr, cursor_row, start_char, cursor_row, #line, { '' })
+    return line:sub(cursor_col + 1)
+  end
+
+  --- @param suffix? string
+  local function apply_snippet_and_command(suffix)
+    if expand_snippet then
+      apply_snippet(completion_item, suffix)
+    end
+
+    local command = completion_item.command
+    if command then
+      local fn = client.commands[command.command] or lsp.commands[command.command]
+      if fn then
+        fn(command, { bufnr = bufnr, client_id = client_id })
+      else
+        local command_provider = client.server_capabilities.executeCommandProvider or {}
+        local server_commands = command_provider.commands or {}
+        if vim.tbl_contains(server_commands, command.command) then
+          client.request(ms.workspace_executeCommand, {
+            command = command.command,
+            arguments = command.arguments,
+          }, function() end, bufnr)
+        else
+          vim.notify(
+            'Command not supported on client or server: ' .. command.command,
+            vim.log.levels.WARN
+          )
+        end
+      end
+    end
+  end
+
+  if completion_item.additionalTextEdits and next(completion_item.additionalTextEdits) then
+    local suffix = clear_word()
+    lsp.util.apply_text_edits(completion_item.additionalTextEdits, bufnr, offset_encoding)
+    apply_snippet_and_command(suffix)
+  elseif resolve_provider and type(completion_item) == 'table' then
+    local changedtick = vim.b[bufnr].changedtick
+
+    --- @param result lsp.CompletionItem
+    client.request(ms.completionItem_resolve, completion_item, function(err, result)
+      if changedtick ~= vim.b[bufnr].changedtick then
+        return
+      end
+
+      local suffix = clear_word()
+      if err then
+        vim.notify_once(err.message, vim.log.levels.WARN)
+      elseif result and result.additionalTextEdits then
+        lsp.util.apply_text_edits(result.additionalTextEdits, bufnr, offset_encoding)
+        if result.command then
+          completion_item.command = result.command
+        end
+      end
+
+      apply_snippet_and_command(suffix)
+    end, bufnr)
+  else
+    local suffix = clear_word()
+    apply_snippet_and_command(suffix)
+  end
+end
+
+--- Trigger LSP completion in the current buffer.
+function M.trigger_completion()
```

Comment:
This function is mostly called in "scheduled" fashion (via `vim.shedule()` or after `vim.schedule_wrap()`). Usually it should be possible to do single `schedule_wrap()` after defining function instead of all other scheduling in place.